sample/CMakeLists.txt

```diff
@@ -0,0 +1,42 @@
+cmake_minimum_required(VERSION 3.8.0)
+
+project(pcg-samples)
+
+include(GNUInstallDirs)
+
+if (NOT TARGET pcg::pcg)
+  find_package(pcg CONFIG REQUIRED)
+endif()
```

Comment:
Not sure why this is needed? 😕

Reply:
This is needed because all the tests are linking against `pcg::pcg`. `pcg::pcg` is a target in CMake.

There are two possibilities:
1. You have done a super-build; `pcg::pcg` exists as a target in CMake's cache. This is because the `CMakeLists.txt` file in the root of the source tree parses the directories accordingly,
2. You have not done a super-build; in this case, you should have installed `pcg` under one of the default directories CMake searches the libraries for. That's why Line 8 is explicitly asking for the CMake-installed `pcg` package (i.e., `find_package(pcg CONFIG REQUIRED)`).

Reply:
Yeah I sort of realised shortly after that you may have been going for being able to build the tests standalone as well as from the root. I've not seen any library do that before and unsure of the practicallity of that 🤷‍♂ .

Reply:
Well... As we discussed back then, it seemed to be a nice approach (credits go to Daniel Pfeifer, by the way). Using CMake as a dependency manager (at least, in the context of the library itself) eases a lot of pain in advance. As a consumer of `pcg` (or, any other library) I should not be forced to remember any "options" to turn on/off testing and other features.

Reply:
Now if only we could get *everyone* to do that... 😅

Reply:
That's why I was trying this way of wrapping the library in CMake.

Take a look at [`cereal`](https://github.com/uscilab/cereal), for instance. I know that it's a bit of an extreme, but to understand what's going on, you need to dig into the `CMakeLists.txt` file. There are hard-coded compiler options, global definitions, etc. If you are only interested in installing the library, you have to write `cmake -D JUST_INSTALL_CEREAL=ON <source-tree>`. That could be fine for you or somebody else who is willing to only install the library. But what happens if you are a consumer of `cereal` and add it as a target to your CMake project? You want to simply write

```cmake
add_subdirectory(external/cereal/lib) # or, whatever
```

You should not care about the library's tests, compiler switches, options, etc. With the current solution of `cereal`, whenever you add it to a CMake project, it's tests are built! It adds to the compilation time of your own project. The situation is similar in other projects such as `openblas`, `zeromq`, etc. But does that mean they are correct? Does that mean that their way of doing things is better? I am not so sure...

I don't know... That's my personal preference. I could as well be wrong.

But definitely we can simplify the script and have 3 CMake projects: one for `tests`, one for `samples` and one for the root, i.e., the `pcg` target. If you are interested in the library, build and install the `pcg` project. If you want to test, after building and installing the `pcg` project, build the `tests` project. The same is true for the `samples` project. This way, we do not have the super build, and hence no need to check for the existence of the `pcg::pcg` target.

Reply:
My line of reasoning is that most people are lazy and don't read docs until they hit a problem (I'm guilty of this). So I would imagine most people would use the root CMakeLists for all use cases and then wonder how to not build tests by opening it up and seeing no option for it and then 
*finally* go look at the docs.

The libraries that I've used that have good CMake usage only build tests if it's the main project (which is a sensible thing to do IMO) which is why I made this comment, https://github.com/imneme/pcg-cpp/pull/44#discussion_r373999215.

Examples:
https://github.com/CLIUtils/CLI11/blob/master/CMakeLists.txt#L61
https://github.com/fmtlib/fmt/blob/master/CMakeLists.txt#L43
https://github.com/foonathan/type_safe/blob/master/CMakeLists.txt#L133

Reply:
> So I would imagine most people would use the root CMakeLists for all use cases and then wonder how to not build tests by opening it up and seeing no option for it and then finally go look at the docs.

Fine. But what is your proposal? Under this PR, we have discussed these (or similar) issues. Who is our target audience?

1. The impatient will just go find the `include` directory, use the 3 files. Done.
2. The developers (hopefully) check the "Installation" section under the landing page (`README.md`) and see that the root has the super-build functionality.
3. Users relying on their package managers will simply issue their package manager's installation subcommand.

First category above is not the target audience, I believe. Regardless of what we do here, they will go fetch the files. Third group will not have any problems at all because all they need to write in their CMake projects is `find_package(pcg CONFIG REQUIRED)`. The second group of people are (mostly, I believe) are the most "nerd" of them all. If they know how to track their dependencies in their source trees using `git submodule`, for instance, they should be reading the Installation section in the README, which should take 1 minute.

That said, what do _we_ want? If you say "keep the root of the source tree as the CMake project for `pcg` and make `samples` and `tests` separate CMake projects that writes `find_package(pcg CONFIG REQUIRED)`," I am happy to assist. Then, we will have 3 projects and no options, and I am happy.

If you would like to go with the "options" route, I am sorry. Then, please do however you like. I do not like that route, and I will not work on something I do not like. Similar to the examples you have provided, I can provide you counter-examples that prefer this route (or, talks that suggest against the use of options in CMake projects (i.e., directories under which `CMakeLists.txt` exists). The problem with options, which I also agree, is that the names are global. If a user has `PCG_TESTS` variable/option defined in their CMake project, what happens? You are relying on people being rational (or, behaving properly) when you make a decision based on variable names. If we assume they are rational enough, they can as well skim the first couple of lines of the README.

If you definitely need your options for some building/testing/packaging purposes, CMake projects are not the place for it. We have [`cpack`](https://cmake.org/cmake/help/v3.8/manual/cpack.1.html) where we can define configuration options and variables for CMake projects. This solution will let you have the best of both worlds, in my opinion.

Please feel free to have the `43-cmake` branch of [aytekinar/pcg-cpp](https://github.com/aytekinar/pcg-cpp) as a base and introduce your options.

Reply:
As a side note, [this line](https://github.com/CLIUtils/CLI11/blob/706b14fb14444e68ace232eb9a0ef106bf99343b/CMakeLists.txt#L42) and the following two lines (and similar commands in CMake projects) are exactly the ones I am against (or, people giving talks in CppCon or similar venues are against).

CMake projects should define dependencies between files, modules and among different CMake projects. Options and logic around compilers, building, documentation, testing, etc. are not relevant. CMake is _not_ a make tool. It is not a place where we script our logic as in "if it's clang, do this," or "if you want to build docs, do that."

If you follow the talks mentioned in the CMake best practices website you have provided, you will see that people have a `cmake` folder to support this type of logic. All the three libraries you show as examples define some sort of compiler options. I do not want them. As a library writer, you should _not_ decide which compiler options the consumer _has to_ use. Even worse, you should _not_ use global/project-level variables to set those options. On the other hand, if you would like to have your favorite `-Wall -Wpedantic` options for the gcc-family, do so in a `cmake/gcc.cmake` file.

You can check my library (again, I wanted to follow the best practices in these conferences) to have that structure: [`polo`](https://github.com/pologrp/polo). Is it the best library of them all? Definitely not. But my point is that all the build/test logic goes under [`cmake/ctest-gcc.cmake`](https://github.com/pologrp/polo/blob/master/cmake/ctest-gcc.cmake), which handles the warning options, coverage, etc. The same file tells `ctest` to use `gcov` for test coverage and `valgrind` to check against leaks. This is my concern. But as a consumer of `polo`, if you look at the `src` CMake project, you will only see the dependencies and CXX features required for this library. You can happily use it within your CMake project --- I do not make any intrusive definitions that would change the way you build your project depending on `polo`.

Reply:
> Fine. But what is your proposal?

Sorry again for my lack of clarity. I was just explaining my rationale for my original comments and not advocating one way or the other.

> keep the root of the source tree as the CMake project for pcg and ...

That's personally how I would have it just on the mere reasoning being that I find `include` having a CMakeLists a bit weird and unusual. 
If one were to move it out of there, would that not make the samples and tests either forced on you (or only when it's the master project) or forced to do them separately? What would you recommend the approach be in that case without using options?

> As a side note...

That is interesting and didn't notice that option. I don't know why that's an option that's provided to be honest. It's also interesting because `CLIUtils` is the owner of both that library and the git book I linked previously 🤔 .

Reply:
> Sorry again for my lack of clarity. I was just explaining my rationale for my original comments and not advocating one way or the other.

I am not trying to argue with you. I want to know what you or other developers here would 
like to have. This is the first attempt to provide `cmake` support to `pcg`, which also gets rid of the `Makefile`-based approach. The change is breaking. We do not need to think much about breaking changes, because we are already breaking the current flow of the build/test/package process.

My first proposal was to have
```
.
├── samples
│   └── CMakeLists.txt
├── src
│   ├── CMakeLists.txt
│   └── include
└── tests
    └── CMakeLists.txt
```
as `src` is an acronym for "source." Then, this proposal got opposed because it was breaking the mental model of having `src` and `include` under some directory to have implementation and header files, respectively. Moreover, because `src` is used mostly for implementation files and because `pcg` has only header files, some thought `src` could give the wrong impression. Even though I do not agree with those opposing ideas, I can understand the mental model they have, and that's fine.

Then, we can discuss and choose to have something like:
```
.
├── pcg
│   ├── CMakeLists.txt
│   └── include
├── samples
│   └── CMakeLists.txt
└── tests
    └── CMakeLists.txt
```
At least, we have 3 separate CMake projects for 3 different purposes. You can argue that having `pcg/include` and `pcg/CMakeLists.txt` directly under the root (i.e., the source tree) is better, and that's also fine. However, in other communities, there seems to be a shift towards such an indirection of locating "source" files of your package under `src` or similar directories (see, e.g., [this](https://hynek.me/articles/testing-packaging/) for Python projects). The argument around whether such an approach is an overkill for small libraries and such is also going on there. However, it provides _flexibility_ and, more importantly, _isolation_. Depending on my purpose, I can isolate myself in one of the 3 CMake projects.

> That is interesting and didn't notice that option. I don't know why that's an option that's provided to be honest. It's also interesting because `CLIUtils` is the owner of both that library and the git book I linked previously 🤔 .

With all due respect for that author, all I can say is that his approach in `CLIUtils` is not good. As I mentioned before, I would not want any library writer to dictate compiler flags/options for my CMake project. If I am going to include `CLIUtils`, for instance, under my CMake project, I would be forced to use his favourite warning flags! The problem is with trying to use CMake in an imperative way. It should be used in a more declarative way, where you simply build a dependency graph between different projects.

As a side note, it's also interesting when Linus Torvalds [states](http://harmful.cat-v.org/software/c++/linus)

> Quite frankly, even if  the choice of C were to do *nothing* but keep the C++ programmers out, that in itself would be a huge reason to use C.

He is a great person, in my opinion, who made Linux and git a reality of our lives. Yet, this does not mean that his opinion about C++ is correct and valid (perhaps, it is, in a general context).

Reply:
>  some thought src could give the wrong impression.

I would argue that as well.

> Then, we can discuss and choose to have something like:

I misread and thought you were suggesting the other way around earlier i.e. `include/pcg` which is a popular convention. Though that would be a bit silly since all headers are prefixed with `pcg` so that would be superfluous 🤷‍♂️.

Are there any well known/popular (header-only) libraries that use that structure so I can take a look?
My main line of thinking is that if it's not a big deal (which IMO this is - root being for `pcg` or `include` being for `pcg`) do what people expect.